File: typescript/timestep/dbos_config.ts
/** DBOS configuration for Timestep workflows. */

import { DBOS } from '@dbos-inc/dbos-sdk';
import { getPgliteDir } from './app_dir.ts';
import * as path from 'path';

class DBOSContext {
  private config: { name: string; systemDatabaseUrl: string } | null = null;
  private configured = false;
  private launched = false;
  private pgliteServer: any = null;
  private pgliteDb: any = null;

  getConnectionString(): string | undefined {
    if (this.config) {
      return this.config.systemDatabaseUrl;
    }
    return undefined;
  }

  get isConfigured(): boolean {
    return this.configured;
  }

  get isLaunched(): boolean {
    return this.launched;
  }

  setConfig(config: { name: string; systemDatabaseUrl: string }): void {
    this.config = config;
  }

  setConfigured(value: boolean): void {
    this.configured = value;
  }

  setLaunched(value: boolean): void {
    this.launched = value;
  }

  getPgliteServer(): any {
    return this.pgliteServer;
  }

  setPgliteServer(server: any): void {
    this.pgliteServer = server;
  }

  getPgliteDb(): any {
    return this.pgliteDb;
  }

  setPgliteDb(db: any): void {
    this.pgliteDb = db;
  }
}

// Singleton instance
const dbosContext = new DBOSContext();

/**
 * Get the DBOS connection string if configured.
 * 
 * @returns Connection string or undefined if not configured
 */
export function getDBOSConnectionString(): string | undefined {
  return dbosContext.getConnectionString();
}

export interface ConfigureDBOSOptions {
  /** Application name for DBOS (default: "timestep") */
  name?: string;
  /** Optional system database URL. If not provided, uses PG_CONNECTION_URI environment variable, or PGLite */
  systemDatabaseUrl?: string;
}

/**
 * Configure DBOS for Timestep workflows.
 * 
 * Uses PG_CONNECTION_URI for the system database if available,
 * otherwise uses PGLite via socket server (same as RunStateStore).
 * 
 * @param options Configuration options
 */
export async function configureDBOS(options: ConfigureDBOSOptions = {}): Promise<void> {
  const { name = 'timestep', systemDatabaseUrl } = options;
  
  // Get system database URL from parameter, env var, or default to PGLite
  const env = typeof process !== 'undefined' ? process.env : {};
  let dbUrl = systemDatabaseUrl || env['PG_CONNECTION_URI'];
  
  // If no connection string provided, start PGLite socket server
  if (!dbUrl) {
    // Use a dedicated PGLite database for DBOS system database
    const pglitePath = getPgliteDir({ appName: name });
    const dbPath = path.join(pglitePath, 'dbos_system');
    
    // Ensure parent directory exists
    const fs = await import('fs/promises');
    try {
      await fs.mkdir(path.dirname(dbPath), { recursive: true });
    } catch (e: any) {
      // Ignore if directory already exists
      if (e.code !== 'EEXIST') {
        throw e;
      }
    }
    
    // Start PGLite socket server
    const { PGlite } = await import('@electric-sql/pglite');
    const { PGLiteSocketServer } = await import('@electric-sql/pglite-socket');
    const { uuid_ossp } = await import('@electric-sql/pglite/contrib/uuid_ossp');
    
    // Create PGLite instance with uuid-ossp extension (required by DBOS)
    const pgliteDb = new PGlite(dbPath, { 
      dataDir: dbPath,
      extensions: { uuid_ossp }
    });
    await pgliteDb.waitReady;
    dbosContext.setPgliteDb(pgliteDb);
    
    // Use TCP connection for simplicity and compatibility
    // Port 0 means let the OS assign an available port
    const port = 0;
    
    const pgliteServer = new PGLiteSocketServer({
      db: pgliteDb,
      port: port,
      host: '127.0.0.1',
    });
    
    await pgliteServer.start();
    dbosContext.setPgliteServer(pgliteServer);
    
    // Wait for server to initialize (PGLite socket server needs a moment)
    await new Promise(resolve => setTimeout(resolve, 500));
    
    // Get the actual port that was assigned
    const server = (dbosContext.getPgliteServer() as any).server;
    if (!server) {
      throw new Error('Server object not found after start()');
    }
    
    const actualPort = server.address()?.port;
    if (!actualPort) {
      throw new Error('Failed to get port from PGLite socket server');
    }
    
    // Test the connection before using it
    // Note: PGLite socket server only supports one connection at a time,
    // so we test and immediately close to ensure DBOS can connect
    try {
      const pg = await import('pg');
      const { Client } = pg;
      const testClient = new Client({
        host: '127.0.0.1',
        port: actualPort,
        user: 'postgres',
        password: 'postgres',
        database: 'postgres',
        ssl: false,
        connectionTimeoutMillis: 10000,
      });
      await testClient.connect();
      await testClient.query('SELECT 1');
      await testClient.end();
      // Wait for connection to fully close before DBOS connects
      // PGLite socket server only supports one connection at a time
      await new Promise(resolve => setTimeout(resolve, 500));
    } catch (e: any) {
      throw new Error(`PGLite socket server connection test failed: ${e.message}`);
    }
    
    // Construct PostgreSQL connection string pointing to the socket server
    dbUrl = `postgresql://postgres:postgres@127.0.0.1:${actualPort}/postgres?sslmode=disable`;
    console.log(`PGLite socket server listening on port ${actualPort}`);
    console.log(`DBOS will use connection string: postgresql://postgres:postgres@127.0.0.1:${actualPort}/postgres?sslmode=disable`);
  }
  
  // DBOS will use the same database but different schema (dbos schema)
  const config = {
    name,
    systemDatabaseUrl: dbUrl,
  };
  console.log(`Setting DBOS config with systemDatabaseUrl: ${config.systemDatabaseUrl}`);
  DBOS.setConfig(config);
  dbosContext.setConfig(config);
  dbosContext.setConfigured(true);
}

/**
 * Ensure DBOS is launched. Safe to call multiple times.
 * 
 * This should be called before using any DBOS workflows.
 */
export async function ensureDBOSLaunched(): Promise<void> {
  if (!dbosContext.isConfigured) {
    await configureDBOS();
  }
  
  if (!dbosContext.isLaunched) {
    // Verify the config before launching
    const connectionString = dbosContext.getConnectionString();
    console.log(`DBOS config before launch - connection string: ${connectionString}`);
    if (!connectionString) {
      throw new Error('DBOS connection string is not set. Call configureDBOS() first.');
    }
    
    // Ensure config is set in DBOS (in case it was cleared or reset)
    const config = dbosContext.getConnectionString();
    if (config) {
      DBOS.setConfig({
        name: 'timestep',
        systemDatabaseUrl: config,
      });
      console.log(`Re-set DBOS config with connection string: ${config}`);
    }
    
    console.log('Calling DBOS.launch()...');
    try {
      await DBOS.launch();
      console.log('DBOS.launch() completed');
      dbosContext.setLaunched(true);
    } catch (error: any) {
      console.error('DBOS.launch() failed:', error);
      // Log the actual config DBOS is trying to use
      const currentConfig = dbosContext.getConnectionString();
      console.error(`DBOS was trying to connect to: ${currentConfig}`);
      throw error;
    }
  }
}

/**
 * Clean up PGLite socket server and database connections.
 * Call this when shutting down the application.
 */
export async function cleanupDBOS(): Promise<void> {
  const pgliteServer = dbosContext.getPgliteServer();
  if (pgliteServer) {
    await pgliteServer.stop();
    dbosContext.setPgliteServer(null);
  }
  const pgliteDb = dbosContext.getPgliteDb();
  if (pgliteDb) {
    await pgliteDb.close();
    dbosContext.setPgliteDb(null);
  }
  dbosContext.setConfigured(false);
  dbosContext.setLaunched(false);
}

File: python/timestep/dbos_config.py
"""DBOS configuration for Timestep workflows."""

import os
import sys
from pathlib import Path
from typing import Optional
from dbos import DBOS, DBOSConfig
from .app_dir import get_pglite_dir

# Global state for PGLite socket server
_pglite_sidecar = None
_pglite_socket_server_info = None


def configure_dbos(
    name: str = "timestep",
    system_database_url: Optional[str] = None
) -> None:
    """
    Configure DBOS for Timestep workflows.
    
    Uses PG_CONNECTION_URI for the system database if available,
    otherwise will use PGLite via socket server (started in ensure_dbos_launched).
    
    Args:
        name: Application name for DBOS (default: "timestep")
        system_database_url: Optional system database URL. If not provided,
            uses PG_CONNECTION_URI environment variable, or PGLite
    """
    global _pglite_socket_server_info
    
    # Get system database URL from parameter, env var, or default to PGLite
    db_url = system_database_url or os.environ.get("PG_CONNECTION_URI")
    
    # If no connection string provided, prepare for PGLite socket server
    if not db_url:
        # Use a dedicated PGLite database for DBOS system database
        pglite_path = get_pglite_dir()
        db_path = pglite_path / 'dbos_system'
        
        # Use Unix socket for better performance (or TCP if on Windows)
        is_windows = sys.platform == 'win32'
        socket_path = None if is_windows else str(db_path.parent / '.s.PGSQL.5432')
        port = 5432 if is_windows else 0
        
        # Store socket server info for later startup
        _pglite_socket_server_info = {
            'path': socket_path,
            'port': port,
            'host': '127.0.0.1',
            'db_path': db_path
        }
        
        # Placeholder - will be updated in ensure_dbos_launched
        db_url = "postgresql://postgres:postgres@127.0.0.1:5432/postgres?sslmode=disable"
    
    # DBOS will use the same database but different schema (dbos schema)
    config: DBOSConfig = {
        "name": name,
        "system_database_url": db_url,
    }
    
    DBOS(config=config)


_dbos_configured = False
_dbos_launched = False


async def ensure_dbos_launched() -> None:
    """
    Ensure DBOS is configured and launched. Safe to call multiple times.
    
    This should be called before using any DBOS workflows.
    """
    global _dbos_configured, _dbos_launched, _pglite_sidecar, _pglite_socket_server_info
    
    if not _dbos_configured:
        configure_dbos()
        _dbos_configured = True
    
    # If using PGLite, start socket server before launching DBOS
    if _pglite_socket_server_info:
        # Import sidecar client
        from .pglite_sidecar.client import PGliteSidecarClient
        
        # Create or reuse sidecar client
        if _pglite_sidecar is None:
            _pglite_sidecar = PGliteSidecarClient(str(_pglite_socket_server_info['db_path']))
            await _pglite_sidecar.start()
        
        # Start socket server and get connection string
        try:
            server_info = await _pglite_sidecar.start_socket_server(
                port=_pglite_socket_server_info['port'],
                host=_pglite_socket_server_info['host'],
                path=_pglite_socket_server_info['path']
            )
            connection_string = server_info.get('connectionString')
            if not connection_string:
                raise RuntimeError("Failed to get connection string from PGLite socket server")
            
            # Reconfigure DBOS with the actual connection string
            config: DBOSConfig = {
                "name": "timestep",  # Use default name
                "system_database_url": connection_string,
            }
            DBOS(config=config)
        except Exception as e:
            raise RuntimeError(f"Failed to start PGLite socket server for DBOS: {e}")
    
    if not _dbos_launched:
        DBOS.launch()
        _dbos_launched = True

